pkg/alluxiocluster/controller.go

```diff
@@ -32,88 +33,92 @@ type AlluxioClusterReconciler struct {
 	}
 
 	alluxioCluster := &alluxiov1alpha1.AlluxioCluster{}
-	ctx.AlluxioCluster = alluxioCluster
-	if err := r.Get(context, req.NamespacedName, alluxioCluster); err != nil {
-		if errors.IsNotFound(err) {
-			logger.Infof("Alluxio cluster %s not found. It is being deleted or already deleted.", req.NamespacedName.String())
-		} else {
-			logger.Errorf("Failed to get Alluxio cluster %s: %v", req.NamespacedName.String(), err)
-			return ctrl.Result{}, err
-		}
+	ctx.AlluxioClusterer = alluxioCluster
+	if err := GetAlluxioClusterFromK8sApiServer(r, req.NamespacedName, alluxioCluster); err != nil {
+		return ctrl.Result{}, err
 	}
+
 	dataset := &alluxiov1alpha1.Dataset{}
 	datasetNamespacedName := types.NamespacedName{
-		Name:      alluxioCluster.Spec.Dataset,
+		Name:      *alluxioCluster.Spec.Dataset,
 		Namespace: req.Namespace,
 	}
-	ctx.Dataset = dataset
-	if err := r.Get(context, datasetNamespacedName, dataset); err != nil {
+	ctx.Datasetter = dataset
+	if err := datasetPkg.GetDatasetFromK8sApiServer(r, datasetNamespacedName, dataset); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := DeleteClusterIfNeeded(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := CreateAlluxioClusterIfNeeded(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return UpdateStatus(ctx)
+}
+
+func GetAlluxioClusterFromK8sApiServer(r client.Reader, namespacedName types.NamespacedName, alluxioCluster *alluxiov1alpha1.AlluxioCluster) error {
+	if err := r.Get(context.TODO(), namespacedName, alluxioCluster); err != nil {
 		if errors.IsNotFound(err) {
-			logger.Infof("Dataset %s not found. It is deleted or hasn't been created yet.", alluxioCluster.Spec.Dataset)
-			dataset.Status.Phase = alluxiov1alpha1.DatasetPhaseNotExist
+			logger.Infof("Alluxio cluster %s not found. It is being deleted or already deleted.", namespacedName.String())
 		} else {
-			logger.Errorf("Failed to get Dataset %s: %v", req.NamespacedName.String(), err)
-			return ctrl.Result{}, err
+			logger.Errorf("Failed to get Alluxio cluster %s: %v", namespacedName.String(), err)
+			return err
 		}
 	}
+	return nil
+}
 
-	if alluxioCluster.DeletionTimestamp != nil {
-		if err := deleteConfYamlFileIfExist(ctx.NamespacedName); err != nil {
-			return ctrl.Result{}, err
+func DeleteClusterIfNeeded(ctx *AlluxioClusterReconcileReqCtx) error {
+	if ctx.AlluxioClusterer.IsDeleted() {
+		if err := DeleteConfYamlFileIfExist(ctx.NamespacedName); err != nil {
+			return err
 		}
 		if err := DeleteAlluxioClusterIfExist(ctx.NamespacedName); err != nil {
-			return ctrl.Result{}, err
+			return err
 		}
-		if dataset.Status.Phase != alluxiov1alpha1.DatasetPhaseNotExist {
-			ctx.Dataset.Status.Phase = alluxiov1alpha1.DatasetPhasePending
-			ctx.Dataset.Status.BoundedAlluxioCluster = ""
+		if *ctx.Datasetter.GetStatus().Phase != alluxiov1alpha1.DatasetPhaseNotExist {
+			pending := alluxiov1alpha1.DatasetPhasePending
+			ctx.Datasetter.GetStatus().Phase = &pending
```

Comment:
```suggestion
		if ctx.Datasetter.GetStatus().Phase != alluxiov1alpha1.DatasetPhaseNotExist {
			ctx.Datasetter.GetStatus().Phase = alluxiov1alpha1.DatasetPhasePending
```
if `Phase` is a string type rather than *string type